Convert options file to YAML and use uWSGI by default

jmchilton opened the pull request:
## What's In It?

The default sample configuration for Galaxy is now ``config/galaxy.yml.sample`` instead of ``config/galaxy.ini.sample``. If a ``galaxy.yml`` configuration is found - ``run.sh`` will start with uWSGI instead of paste by default. This means all new Galaxy instances will target uWSGI by default.

Beyond just swapping the configuration format, this PR provides important new tooling and documentation for swapping.

- ``make config-convert-dry-run`` - will demonstrate how to convert an existing ``config/galaxy.ini`` into a ``config/galaxy.yml``. 
- ``make config-convert`` - will actually perform this configuration.
- ``config-validate`` - will validate a ``galaxy.yml`` file for formal correctness using pykwalify.
- ``config-lint`` - will lint a ``galaxy.yml`` file and warn of various potential issues.
- The admin section of the documentation now features RST documentation for the Galaxy configuration options.

### Backward Compatibility 

Existing galaxy.ini should continue to work just fine and will still cause Galaxy to start a paster server when coming from ``run.sh`` - so this should be effectively completely backward compatible for existing admins.

## Adding Configuration Options to Galaxy

The sample configuration file still exists, but the file ``lib/galaxy/webapps/galaxy/config_schema.yml`` containing the specification should now be considered the canonical documentation for a Galaxy configuration option. The sample YAML file and the RST documentation can be automatically re-generated from that source using the Make targets ``config-rebuild-sample`` and ``config-rebuild-rst`` respectively.

## Conversion and Linting Galaxy Parameters

The following ``dict`` describes some of the basic handling of converting and linting Galaxy parameters - hopefully it makes a little sense (there are test cases if you want to dig in deeper).

```python
OPTION_ACTIONS = {
    'use_beaker_session': _DeprecatedAndDroppedAction(),
    'session_type': _DeprecatedAndDroppedAction(),
    'session_data_dir': _DeprecatedAndDroppedAction(),
    'session_key': _DeprecatedAndDroppedAction(),
    'session_secret': _DeprecatedAndDroppedAction(),
    'paste.app_factory': _PasteAppFactoryAction(),
    'filter-with': _HandleFilterWithAction(),
    'debug': _ProductionUnsafe(True),
    'serve_xss_vulnerable_mimetypes': _ProductionUnsafe(True),
    'use_printdebug': _ProductionUnsafe(True),
    'use_interactive': _ProductionUnsafe(True),
    'id_secret': _ProductionUnsafe('USING THE DEFAULT IS NOT SECURE!'),
    'master_api_key': _ProductionUnsafe('changethis'),
    'external_service_type_config_file': _DeprecatedAndDroppedAction(),
    'external_service_type_path': _DeprecatedAndDroppedAction(),
    'enable_sequencer_communication': _DeprecatedAndDroppedAction(),
    'run_workflow_toolform_upgrade': _DeprecatedAndDroppedAction(),
    # Next 4 were from library search which is no longer available.
    'enable_lucene_library_search': _DeprecatedAndDroppedAction(),
    'fulltext_max_size': _DeprecatedAndDroppedAction(),
    'fulltext_noindex_filetypes': _DeprecatedAndDroppedAction(),
    'fulltext_url': _DeprecatedAndDroppedAction(),
    'enable_legacy_sample_tracking_api': _DeprecatedAction(),
    'enable_new_user_preferences': _DeprecatedAndDroppedAction(),
    'force_beta_workflow_scheduled_for_collections': _DeprecatedAction(),
    'force_beta_workflow_scheduled_min_steps': _DeprecatedAction(),
    'history_local_serial_workflow_scheduling': _ProductionPerformance(),
    'allow_library_path_paste': _RenameAction("allow_path_paste"),
    'trust_ipython_notebook_conversion': _RenameAction("trust_jupyter_notebook_conversion"),
    'enable_beta_tool_command_isolation': _DeprecatedAndDroppedAction(),
    'single_user': _ProductionUnsafe(True),
    'tool_submission_burst_threads': _ProductionPerformance(),
    'tool_submission_burst_at': _ProductionPerformance(),
    'toolform_upgrade': _DeprecatedAndDroppedAction(),
}
```

If other options should have custom conversion or special warning/info during linting let me know and I'd be happy to add it.

## A Note on Reviewing this PR

I'd encourage review of this PR via commit - at least as it pertains to the options file. I've tried to automate the conversion of the ini sample file into a YAML configuration specification (via kwalify and the infrastructure setup in #3179). I've done this step-by-step so you can see the manual edits of the content versus the automated conversion steps in separate commits. I've also done this in such a way that ``git log --follow lib/galaxy/webapps/galaxy/config_schema.yml`` preserves the history of the ``config/galaxy.ini.sample``.

This PR is broken into a few different chunks. The commits prefixed with ``config_manage.py`` are fixes to the infrastructure for YAML parsing adding in #3719. The commits prefixed with ``convert_yaml`` are the actual conversion process. And finally the commits prefixed with ``yaml_config`` is the generation of new artifact, Makefile targets, etc... for the new configuration specification. 

## TODO

- [x] Rework amqp option parsing to allow specification via YAML.
- [x] Document logging parsing to allow specification via YAML. @natefoo
- [x] Freeze 17.09 documentation and add links to it to dev documentation on pages that heavily mention Galaxy options. (xref #5120, https://github.com/galaxyproject/galaxy/pull/5105/commits/d7a91082302daec43df3ddae1f6c5c40d29f339a)
- [x] Fill out new RST page on Galaxy options - right now it is an index but it should have an introduction and probably some discussion of migrating. Is there a good place to start on the hub for generating material for this? @jmchilton
- [x] Improve list of unsafe, beta, and dropped configuration options to improve new configuration file linter. @jmchilton
- [x] ~Create a Github issue outlining further work - including allowing other configurations to be embedded in ``galaxy.yml``, adding more annotations to ``config_schema.yml`` to improve documentation, and pre-processing the type information to simply ``config.py`` considerably.~ https://github.com/galaxyproject/galaxy/issues/5148 
- [x] Update nginx configuration related proxy prefix handling (still references ini) - see nginx.md. @natefoo
- [x] Update apche configuration related proxy prefix handling (still references ini) - see apache.md. @natefoo
- [x] Update description of "servers" in jobs.md (still references ini). @natefoo
- [x] Re-write scaling.md to eliminate references to ini and reflect newer uWSGI by default. @natefoo
- [x] First first setup CI test - still broken.

There is a corresponding WIP Hub PR here (https://github.com/galaxyproject/galaxy-hub/pull/345).

natefoo committed:
Begin config schema for Galaxy (just the new not-in-galaxy.ini logging
option for now, the rest will be autogenerated from the sample)

natefoo committed:
WIP admin documentation updates

jmchilton committed:
yaml_config - Add entry points (e.g. Makefile) for dealing with Galaxy YAML configs.

jmchilton committed:
config_manage.py - allow printing comment header on sample description files.

Generated from the desc field in the app schema.

jmchilton committed:
config_manage.py - fix sample file generation for large default values.

The various messages in the Galaxy config were being split across lines and generating invalid YAML.

jmchilton committed:
yaml_config - Generate Galaxy yaml sample.

jmchilton committed:
config_manage.py - hack uWSGI section of sample file generation.

Because uWSGI seems really bad at parsing fairly simple, entirely valid YAML.

jmchilton committed:
yaml_config - Add RST docs for Galaxy options.

Automatically generated from the new specification.

jmchilton committed:
config_manage.py - improve default value formatting in RST.

jmchilton committed:
Random fix for tool filters.

jmchilton committed:
convert_yaml - setup copy of galaxy.ini.sample for conversion to work on.

jmchilton committed:
convert_yaml - script to convert galaxy.ini.sample to config_schema.yml.

jmchilton committed:
convert_yaml - make sample a schema.

jmchilton committed:
yaml_config - update code base for converting galaxy.ini.sample to YAML.

jmchilton committed:
Refactor galaxy.ini.sample config comments toward per-option docs.

jmchilton committed:
convert_yaml - cleanup noise in conversion's copy of galaxy.ini.sample before conversion.

jmchilton committed:
Fix ignoring ".sample" in newish path module.

Required to correctly detect the extension of sample configuration files.

jmchilton committed:
yaml_config - update admin docs for switch to YAML config.

bgruening commented on doc/source/admin/authentication.md:
````diff
@@ -75,29 +75,30 @@ How to set up this config is presented here.
 
 ```
-track_jobs_in_database = True
+track_jobs_in_database true
````
colon missing?

jmchilton replied:
Yup - I'll get it on a rebase. Thanks.

jmchilton committed:
yaml_config - change test wrapper to work with either ini or yml in container.

jmchilton committed:
convert_yaml - run conversion to YAML.

jmchilton committed:
convert_yaml - remove tmp directory for converting yaml.

jmchilton committed:
config_manage.py - fix to generate YAML booleans in RST docs.

The new default - might as well use that representation.

jmchilton committed:
Don't read global configuration for reports - it doesn't use it.

jmchilton committed:
yaml_config - Don't parse YAML configurations as ini files.

jmchilton committed:
yaml_config - Try to fix first startup CI job for uWSGI default.

jmchilton committed:
Axe special amqp config handling not used by modern messaging stuff.

It wasn't compatible with YAML config and looks unused as far as I can tell.

xref https://github.com/galaxyproject/galaxy/commit/6b92bf6eb8ca8a5114fc2ecf7fda64c011524b9f

natefoo committed:
Don't use uWSGI master mode by default. Fixes Ctrl+C on macOS and is
really only important in production configs where you should be using a
config file anyway.

jmchilton commented:
@natefoo The first startup CI test is broken - any ideas?

```
executing: .venv/bin/uwsgi --module 'galaxy.webapps.galaxy.buildapp:uwsgi_app()' --virtualenv /home/travis/build/galaxyproject/galaxy/.venv --pythonpath lib --master --threads 4 --http localhost:8080 --static-map /static/style=/home/travis/build/galaxyproject/galaxy/static/style/blue --static-map /static=/home/travis/build/galaxyproject/galaxy/static --die-on-term --hook-master-start 'unix_signal:2 gracefully_kill_them_all' --hook-master-start 'unix_signal:15 gracefully_kill_them_all' --enable-threads --py-call-osafterfork  --daemonize2 galaxy.log --safe-pidfile galaxy.pid
/home/travis/build/galaxyproject/galaxy/.venv/bin/python2.7: unrecognized option '--safe-pidfile'
getopt_long() error
cat: galaxy.pid: No such file or directory
.ci/first_startup.sh: line 13: kill: `': not a pid or valid job spec
exit code:1, showing startup log:
cat: galaxy.log: No such file or directory
```

jmchilton committed:
yaml_config - Update Conda FAQ to link out to last .ini variant of docs for those who need them.

natefoo commented:
Looks like it's documented but was never merged to uWSGI 2.0: unbit/uwsgi#1417. I'll open a PR for that, but in the meantime I guess we will need to do this with `--pidfile`, which unfortunately has some issues that `--safe-pidfile` was created to solve.

natefoo commented:
2.0 backport PR'd upstream in unbit/uwsgi#1691, hopefully this will be in 2.0.16. In the meantime, I've switched out `--safe-pidfile` for `--pidfile2`. Without `--safe-pidfile` it means if you accidentally run `sh run.sh --daemon` twice, you can no longer stop the old one with `sh run.sh --stop-daemon`.

natefoo committed:
Drop missing --safe-pidfile uWSGI arg until it is released.

natefoo committed:
Fix the use of --set galaxy_config_file with uWSGI where the app name is
the PasteDeploy default `app:main`.

natefoo commented:
I didn't necessarily mean for that last commit to go here but it's sorta become the "uWSGI post-merge fixes" PR. I can strip it if you'd prefer not to have it mixed in here though.

jmchilton committed:
Fix for None config_file in web stack stuff.

jmchilton committed:
[17.09] Deployer doc fixes.

jmchilton committed:
Update newer deployer config docs for switch to YAML config.

jmchilton commented:
Last night I rebased the new 17.09 docs into this and then modified them for many of the config changes. This morning @natefoo and I agreed to trade some PR reviews and testing for documentation - so I've updated the checklist above with who needs to do what to get this done.

natefoo committed:
WIP admin docs updates for 18.01

jmchilton committed:
Expand number of parameters with custom conversion/linting for Galaxy.

Add functionality and test case for converting parameter names (e.g. from allow_library_path_paste to allow_path_pate) during conversion.

jmchilton commented:
The default config with --daemon launches two uwsgi processes - at least it seems to on my mac. Killing one - the one that write the pid file (or writes the pid file later) doesn't kill the other. I don't know if this is the problem on Travis.

natefoo commented:
What are the two processes? Their pids should be logged on startup.

natefoo committed:
Locate the Galaxy config file when multiple uWSGI --yaml/--json
arguments are used.

martenson commented:
This is marked as WIP so I am pushing it to 18.05.

natefoo committed:
uWSGI: Use --master when daemonizing (required when using http server w/
daemonization) and fix `run.sh --stop-daemon`

natefoo commented:
There are some important 18.01 fixes in here - they should be split out and applied to 18.01 if this is not merged before branching.

natefoo committed:
Major updates to the scaling doc for uWSGI

natefoo committed:
Remove emperor mode reference

natefoo committed:
Add docstring to LOGGING_CONFIG_DEFAULT so that it is part of the
autodoc generated docs, so that it can be linked from statically written
docs.

natefoo committed:
Galaxy config doc corrections

natefoo committed:
Add intersphinx mapping to the Python 2.7 documentation

natefoo committed:
Finish Galaxy logging configuration documentation

natefoo committed:
Add missing generated docs to .gitignore

natefoo committed:
Add loggging configuration to admin docs TOC and move the cluster docs
after the scaling docs

natefoo committed:
Correct the handlers section of the job conf documentation.

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -1,314 +1,570 @@
+However, a better option to managing processes by hand is to use a process manager as documented in the [Starting and
+Stopping](#starting-and-stopping) section.
 
-## Proxy Server
```
@natefoo Are you planning to add the removed instructions for configuring nginx and Apache for uWSGI to `doc/source/admin/special_topics/nginx.md` and `doc/source/admin/special_topics/apache.md` respectively?

natefoo replied:
You probably saw this, but for anyone else wondering, yes, this was addressed in the nginx/Apache docs.

nsoranzo replied:
Yes, thanks!

natefoo committed:
Fix support for proxy prefix under uWSGI.

natefoo committed:
Update nginx documentation for uWSGI

natefoo committed:
Update Apache documentation for uWSGI

natefoo committed:
Correct documentation line endings

nsoranzo commented on doc/source/admin/nginx.md:
```diff
@@ -0,0 +1,462 @@
+- Galaxy runs as the user `galaxy` with primary group `galaxy`
+- Galaxy is served from the hostname `galaxy.example.org`
+
+Throughout the configuration examples in this document, in order to avoid reptition, `#...` is used to denote a location
```
s/reptition/repetition/

nsoranzo commented on doc/source/admin/nginx.md:
```diff
@@ -0,0 +1,463 @@
+  Use the guides referenced in [basic configuration](#basic-configuration) section to configure SSL properly.
+- If your existing nginx configuration contains a line or included config file defining a default server, be sure to
+  disable it by commenting its `server {}` or preventing its inclusion (under Debian-based operating systems, this
+  is done by removing its symlink from `/etc/nginx/sites-enabled`.
```
Missing `)`

nsoranzo commented on doc/source/admin/nginx.md:
```diff
@@ -0,0 +1,463 @@
+- Galaxy runs as the user `galaxy` with primary group `galaxy`
+- Galaxy is served from the hostname `galaxy.example.org`
+
+Throughout the configuration examples in this document, in order to avoid reptition, `#...` is used to denote a location
```
s/reptition/repetition/

nsoranzo commented on doc/source/admin/apache.md:
```diff
@@ -0,0 +1,410 @@
+- Galaxy runs as the user `galaxy` with primary group `galaxy`
+- Galaxy is served from the hostname `galaxy.example.org`
+
+Throughout the configuration examples in this document, in order to avoid reptition, `#...` is used to denote a location
```
s/reptition/repetition/

natefoo committed:
Correct doc typos caught by @nsoranzo

natefoo committed:
Attempt to reduce duplication in the proxy documentation through the use
of includes, move the multi-config-file discussion to a common page.

natefoo commented:
One last thing I want to fix before this is ready to be rebased: the default uWSGI options in the sample config should match the options that `scripts/get_uwsgi_args.py` generates on the command line.

natefoo committed:
Add a tiny bit about uwsgitop

natefoo committed:
Update galaxy.yml.sample uwsgi section from defaults in
get_uwsgi_args.py

natefoo commented on lib/galaxy/webapps/config_manage.py:
```diff
@@ -51,14 +51,84 @@
+    ('master', {
+        'desc': """Enable the master process manager. Disabled by default for maximum compatibility with CTRL+C, but should be enabled for use with --daemon and/or production deployments.""",
+        'default': 'false',
+        'type': 'str',
```
@jmchilton if I set this to `bool` then the values in the sample are pythonic (e.g. `True` and `False`).

natefoo commented:
Ok, ready for rebasing, @jmchilton.

natefoo committed:
Merge remote-tracking branch 'origin/dev' into ini_to_yaml_config

natefoo committed:
Regenerate galaxy.yml.sample and documentation for newly
added/documented config options.

natefoo committed:
Fix lint error

natefoo commented:
@jmchilton says I'm allowed to merge this but I feel a little funny doing it since half the commits are mine.

natefoo commented:
How about this: everything I added look ok to you @jmchilton?

martenson commented:
huge leap forward, thanks boyz!

jmchilton commented:
> How about this: everything I added look ok to you @jmchilton?

The docs you added are so beautiful that I'm glad our project license doesn't prevent me from publishing a book of poetry that is just snippets of your wonderful words!

martenson commented:
Btw is there an alternative for `run.sh --reload`?

natefoo commented:
@jmchilton hrm, this much flattery is not suspicious at all. ;D

@martenson `--py-autoreload`, but I bet it fails, because there is a bug with uwsgi-as-an-exension that I haven't yet fixed or (and it looks like I haven't created an issue for it) where it thinks its application name for re-execution is `python` (because it doesn't expect its executable to be a script).

You can work around that by doing a `pip install --upgrade --no-binary uwsgi --no-cache-dir uwsgi` to install from source.

jmchilton commented:
Might make sense to have ``--reload`` default to paste for now then.

jxtx commented:
![image](https://user-images.githubusercontent.com/79973/35122683-a1d853f2-fc6d-11e7-81a6-447d94ee368b.png)
